src/stdio.jl

```diff
@@ -52,6 +77,36 @@ function watch_stream(rd::IO, name::AbstractString)
     end
 end
 
+function send_stdio(name)
+    if verbose::Bool && !haskey(task_local_storage(), :IJulia_task)
+        task_local_storage(:IJulia_task, "send $name task")
+    end
+    send_stream(name)
+end
+
+send_stdout(t::Timer) = send_stdio("stdout")
+send_stderr(t::Timer) = send_stdio("stderr")
+
+function send_stream(name::AbstractString)
+    buf = bufs[name]
+    if buf.size > 0
+        s = takebuf_string(buf)
+        #if the last char is not valid, maybe only the start of a partially encoded UTF character
+        #was written, so we write it back to the buffer in the hope that the rest of the character will arrive soon
+        last_ind = chr2ind(s, length(s))
+        last_chrstr = s[last_ind:sizeof(s)]
+        if !isvalid(last_chrstr)
+            write(buf, last_chrstr)
+            last_ind -= 1
+        end
+        if last_ind > 0
```

Comment:
I'm thinking something more like:
```jl
d = takebuf_array(buf)
n = num_utf8_trailing(d)
if n > 0
    write(buf, d[end-(n-1):end])
    resize!(d, length(d) - n)
end
s = UTF8String(d)
!isvalid(s) && s = base64encode(d) # fallback: base64-encode non-UTF8 binary data
```
where we define the following function to compute the leftover bytes in O(1) time:
```
function num_utf8_trailing(d::Vector{UInt8})
    i = length(d)
    # find last non-continuation byte in d:
    while i >= 1 && ((d[i] & 0xc0) == 0x80)
        i -= 1
    end
    i < 1 && return 0
    c = d[i]
    # compute number of expected UTF-8 bytes starting at i:
    n = c <= 0x7f ? 1 : c < 0xe0 ? 2 : c < 0xf0 ? 3 : 4
    nend = length(d) + 1 - i # num bytes from i to end
    return nend == n ? 0 : nend
end
```